openedx/core/djangoapps/theming/storage.py

```diff
@@ -0,0 +1,93 @@
+"""
+Comprehensive Theming support for Django's collectstatic functionality.
+See https://docs.djangoproject.com/en/1.8/ref/contrib/staticfiles/
+"""
+import os.path
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured, SuspiciousOperation
+from django.contrib.staticfiles.storage import StaticFilesStorage, CachedFilesMixin
+from django.utils._os import safe_join
+
+
+class ComprehensiveThemingAwareMixin(object):
+    """
+    Mixin for Django storage system to make it aware of the currently-active
+    comprehensive theme, so that it can generate theme-scoped URLs for themed
+    static assets.
+    """
+    def __init__(self, *args, **kwargs):
+        super(ComprehensiveThemingAwareMixin, self).__init__(*args, **kwargs)
+        COMP_THEME_DIR = getattr(settings, "COMP_THEME_DIR", "")  # pylint: disable=invalid-name
+        if not COMP_THEME_DIR:
+            self.theme_location = None
+            return
+
+        if not isinstance(COMP_THEME_DIR, basestring):
+            raise ImproperlyConfigured("Your COMP_THEME_DIR setting must be a string")
+
+        PROJECT_ROOT = getattr(settings, "PROJECT_ROOT", "")  # pylint: disable=invalid-name
+        if PROJECT_ROOT.endswith("cms"):
+            component = "studio"
+        else:
+            component = "lms"
+        self.theme_location = os.path.join(COMP_THEME_DIR, component, "static")
+
+    @property
+    def prefix(self):
+        """
+        This is used by the ComprehensiveThemeFinder in the collection step.
+        """
+        COMP_THEME_DIR = getattr(settings, "COMP_THEME_DIR", "")  # pylint: disable=invalid-name
+        if not COMP_THEME_DIR:
+            return None
+        theme_name = os.path.basename(os.path.normpath(COMP_THEME_DIR))
+        return "themes/{name}/".format(name=theme_name)
+
+    def themed(self, name):
+        """
+        Given a name, return a boolean indicating whether that name exists
+        as a themed asset in the comprehensive theme.
+        """
+        # Nothing can be themed if we don't have a theme location.
+        if not self.theme_location:
+            return False
+
+        try:
+            path = safe_join(self.theme_location, name)
+        except ValueError:
+            raise SuspiciousOperation("Attempted access to '%s' denied." % name)
+        return os.path.exists(os.path.normpath(path))
```

Comment:
I was trying to copy the logic that Django uses as much as possible. `themed()` is a lot like `exists()`, which is [implemented in Django here](https://github.com/django/django/blob/1.8.7/django/core/files/storage.py#L293-L294). `exists()` calls [`path()`](https://github.com/django/django/blob/1.8.7/django/core/files/storage.py#L306-L307), which calls [`safe_join()`](https://github.com/django/django/blob/1.8.7/django/utils/_os.py#L61-L86). It looks like `safe_join()` now raises `SuspiciousOperation` directly, rather than raising a `ValueError` -- this probably changed between Django 1.4 and Django 1.8. I've updated my code to just let the `SuspiciousOperation` raise. It also looks like `normpath()` is no longer called in Django's code, so I've removed it from mine, as well.

We could do some kind of performance testing, but that would delay this merge further, and at this point multiple teams are hoping to leverage this theming feature in future work. Could we do that testing after the merge?